services/rpz.c

```diff
@@ -106,19 +108,24 @@ rpz_trigger_to_string(enum rpz_trigger r)
 /**
  * Get the label that is just before the root label.
  * @param dname: dname to work on
- * @return: pointer to TLD label
+ * @param maxdnamelen: maximum length of the dname
+ * @return: pointer to TLD label, NULL if not found or invalid dname
  */
 static uint8_t*
-get_tld_label(uint8_t* dname)
+get_tld_label(uint8_t* dname, size_t maxdnamelen)
 {
 	uint8_t* prevlab = dname;
+	size_t dnamelen = 0;
 
 	/* only root label */
 	if(*dname == 0)
 		return NULL;
 
 	while(*dname) {
-		dname = dname+*dname+1;
+		dnamelen += ((size_t)*dname)+1;
```

Comment:
The if(dnamelen > maxdnamelen) check needs to be before the `*dname` parts above, in the if(..==0) and while() and in the dnamelen+= statements, you first need to check that the label length byte fits in the maxdnamelen. With an added statement like `if(dnamelen+1 > maxdnamelen) return NULL;` at the start of the routine. I don't think it is needed at the end of the while loop.

Reply:
Actually I think it is also needed before the `*dname` in the != 0 comparison below, an if statement to check if dnamelen+1 > maxdnamelen.

Reply:
Ack on the check at the start of the routine, in case of a maxdnamelen of 0.

For the second check (in the while) I don't think a new `if` is needed, changing the existing one from `if(dnamelen > maxdnamelen)` to `if(dnamelen+1 > maxdnamelen)` should do the trick.

Reply:
See https://github.com/NLnetLabs/unbound/pull/73/commits/ef120738c00c83117060804d734976da898d365f